File: go/candlestickpatterns/patterns/gravestonedoji.go
package patterns

import (
	"zpano/candlestickpatterns/core"
	"zpano/fuzzy"
)

// GravestoneDoji: a one-candle pattern.
//
// Must have:
// - doji body (very small real body relative to high-low range),
// - no or very short lower shadow,
// - upper shadow is not very short.
func GravestoneDoji(cp *core.CandlestickPatterns) float64 {
	if !cp.Enough(1, cp.DojiBody, cp.VeryShortShadow) {
		return 0.0
	}

	b := cp.Bar(1)
	muDoji := cp.MuLess(core.RealBodyLen(b.O, b.C), cp.DojiBody, 1)
	muShortLS := cp.MuLess(core.LowerShadow(b.O, b.L, b.C), cp.VeryShortShadow, 1)
	muLongUS := cp.MuGreater(core.UpperShadow(b.O, b.H, b.C), cp.VeryShortShadow, 1)

	confidence := fuzzy.TProductAll(muDoji, muShortLS, muLongUS)
	return confidence * 100.0
}

File: ts/candlestick-patterns/patterns/gravestone-doji.ts
/** Gravestone Doji pattern (1-candle). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, upperShadow, lowerShadow } from '../core/primitives.ts';

/**
 * Gravestone Doji: a one-candle pattern.
 *
 * Must have:
 * - doji body (very small real body relative to high-low range),
 * - no or very short lower shadow,
 * - upper shadow is not very short.
 *
 * Returns:
 *     Continuous number in [0, 100]. Higher = stronger signal.
 */
export function gravestoneDoji(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(1, cp.dojiBody, cp.veryShortShadow)) return 0.0;

    const b = cp.bar(1);
    const muDoji = cp.muLessCS(realBody(b.o, b.c), cp.dojiBody, 1);
    const muShortLS = cp.muLessCS(lowerShadow(b.o, b.l, b.c), cp.veryShortShadow, 1);
    const muLongUS = cp.muGreaterCS(upperShadow(b.o, b.h, b.c), cp.veryShortShadow, 1);

    return tProductAll(muDoji, muShortLS, muLongUS) * 100.0;
}
